py/dml/serialize.py

```diff
@@ -372,80 +393,158 @@ def type_signature(dmltype, is_for_serialization):
+                                error_context=None):
+    targets = tuple(targets)
+    imm_attr_decl, imm_attr_ref = declare_variable(
+        site, "_imm_attr", attr_value_t)
+    statements = [imm_attr_decl]
+    for (i, target_info) in enumerate(targets):
+        index = ctree.mkIntegerConstant(site, i, False)
+        sim_attr_list_item = apply_c_fun(site, "SIM_attr_list_item",
+            [val_attr, index], attr_value_t)
+        imm_set = ctree.mkAssignStatement(
+            site, imm_attr_ref,
+            ctree.ExpressionInitializer(sim_attr_list_item))
+        statements.append(imm_set)
+        if target_info is not None:
+            (target, typ) = target_info
+            if typ is not None:
+                def sub_error_out(exc, msg):
+                    sub_error_context = (
+                        f" during deserialization of list item {i}"
+                        + f' during {error_context}' * bool(error_context))
+                    return error_out_at_index(i,
+                                              exc,
+                                              msg + sub_error_context
+                                              if msg is not None else msg)
+                sub_deserialize = deserialize(typ, imm_attr_ref, target,
+                                              sub_error_out)
+            else:
+                sub_deserialize = ctree.mkAssignStatement(
+                    site, target, ctree.ExpressionInitializer(imm_attr_ref))
+            statements.append(sub_deserialize)
+        else:
+            is_not_nil_expr = expr.mkLit(site, '!SIM_attr_is_nil(_imm_attr)',
+                                         TBool())
+            statements.append(ctree.mkIf(
+                site, is_not_nil_expr,
+                ctree.mkCompound(
+                    site,
+                    error_out_at_index(
+                        i,
+                        'Sim_Set_Illegal_Type',
+                        f'expected list item {i} to be nil'
+                        + f' during {error_context}' * bool(error_context)
+                    ))))
+    deserialization = ctree.mkCompound(site, statements)
+    attr_is_list = apply_c_fun(site, "SIM_attr_is_list", [val_attr], TBool())
+
+    attr_list_size = apply_c_fun(site, "SIM_attr_list_size", [val_attr],
+                                 TInt(32, False))
+    attr_list_size_check = ctree.mkEquals(
+        site, attr_list_size,
+        ctree.mkIntegerConstant(site, len(targets), False))
+
+    is_list_expr = ctree.mkAnd(site, attr_is_list, attr_list_size_check)
+    ctree.mkIf(
+        site, is_list_expr, deserialization,
+        ctree.mkCompound(
+            site,
+            error_out_at_index(
+                0, # TODO: None instead?
+                'Sim_Set_Illegal_Type',
+                f'expected list of size {len(targets)}'
+                + f' during {error_context}' * bool(error_context))
+            )
+    ).toc()
```

Comment:
Delegating `.toc` to caller would make this function pure

Reply:
See other comment.